### PR TITLE
feat(delete): Add delete by query

### PR DESCRIPTION
**Pull request by ElPicador**

### CODE REVIEW

File: algoliasearch-common/src/main/java/com/algolia/search/Index.java [hunk 1]
```diff
@@ -1433,44 +1433,74 @@ default Task batchRules(@Nonnull List<Rule> rules, @Nonnull RequestOptions reque
...
+   * @param requestOptions Options to pass to this request
+   * @throws AlgoliaException
+   */
+  default Task deleteBy(@Nonnull Query query, @Nonnull RequestOptions requestOptions) throws AlgoliaException {
+    return getApiClient().deleteBy(getName(), query, requestOptions);
+  }
+
+  /**
    * Delete records matching a query, internally uses browse
+   * Deprecated use deleteBy
```
Comment (clement-leprovost) [inline algoliasearch-common/src/main/java/com/algolia/search/Index.java]: Doesn't Javadoc have a `@deprecated` tag? I know we already have an annotation, but if we put something in the comment, why not use it?

File: algoliasearch-common/src/main/java/com/algolia/search/APIClient.java [hunk 2]
```diff
@@ -968,6 +968,20 @@ void deleteByQuery(String indexName, Query query, int batchSize, RequestOptions
...
+  Task deleteBy(String indexName, Query query, RequestOptions requestOptions) throws AlgoliaException {
+    query = query == null ? new Query() : query;
+    AlgoliaRequest<Task> algoliaRequest = new AlgoliaRequest<>(
+      HttpMethod.POST,
+      false,
+      Arrays.asList("1", "indexes", indexName, "deleteByQuery"),
+      requestOptions,
+      Task.class
+    )
+      .setData(query);
```
Comment (clement-leprovost) [inline algoliasearch-common/src/main/java/com/algolia/search/APIClient.java:980]: What will `setData` do? As per the [REST API Reference](https://www.algolia.com/doc/rest-api/search/#delete-by-query), the API expects a JSON object with a `params` attribute containing all parameters as a URL-encoded string.

File: algoliasearch-common/src/test/java/com/algolia/search/integration/sync/SyncDeleteByTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+  @SuppressWarnings("deprecation")
+  @Test
+  public void deleteByQuery() throws AlgoliaException {
+    Index<AlgoliaObject> index = client.initIndex("index1", AlgoliaObject.class);
+
+    List<AlgoliaObject> objects = IntStream.rangeClosed(1, 10).mapToObj(i -> new AlgoliaObject("name" + i, i)).collect(Collectors.toList());
+    index.addObjects(objects).waitForCompletion();
+
+    index.deleteByQuery(new Query(""));
```
Comment (clement-leprovost) [inline algoliasearch-common/src/test/java/com/algolia/search/integration/sync/SyncDeleteByTest.java:45]: Shouldn't you wait for completion here? Delete by query is asynchronous as all write operations; you must wait on the task ID.

Comment (ElPicador) [reply algoliasearch-common/src/test/java/com/algolia/search/integration/sync/SyncDeleteByTest.java:45]: It's the old deleteByQuery, so no tasks here

File: algoliasearch-common/src/main/java/com/algolia/search/Index.java (hunk 1 above)
Comment (ElPicador) [reply algoliasearch-common/src/main/java/com/algolia/search/Index.java]: Yep, adding it.

File: algoliasearch-common/src/main/java/com/algolia/search/APIClient.java (hunk 2 above)
Comment (ElPicador) [reply algoliasearch-common/src/main/java/com/algolia/search/APIClient.java:980]: `search` also works with a plain JSON objects, does it works here also?

File: algoliasearch-common/src/test/java/com/algolia/search/integration/sync/SyncDeleteByTest.java (hunk 3 above)
Comment (ElPicador) [reply algoliasearch-common/src/test/java/com/algolia/search/integration/sync/SyncDeleteByTest.java:45]: you are right

File: algoliasearch-common/src/main/java/com/algolia/search/APIClient.java (hunk 2 above)
Comment (clement-leprovost) [reply algoliasearch-common/src/main/java/com/algolia/search/APIClient.java:980]: Looking at the code, it seems so, but it's not documented... :/

**Review by clement-leprovost — APPROVED** [review]
LGTM 👍

**Pull request merged**